### ISSUE TITLE
Hand craft a couple of critical messages and run tests on the serialized size

### ISSUE THREAD

**Issue by LorenzMeier** [mavlink/v2_playground#1]
@liamstask I want to run this test on:

  * ATTITUDE_QUATERNION
  * GLOBAL_POSITION_INT
  * MISSION_ITEM
  * VEHICLE_STATUS

And also on these:

https://docs.google.com/spreadsheets/d/1c5Vd-kzI7SOdKRk_d_Ri04z0xFR21Zg5ZMHgyiKkpUA/edit#gid=1551144504

**Comment by liamstask** [comment]
@LorenzMeier - what's your goal with this test? to identify the smallest encoding across a variety of msg formats? if so, do you have a list of the formats you're interest in? might also be helpful to measure encode/decode time, memory usage, etc.

**Comment by LorenzMeier** [comment]
Yes, those quantities would be interesting as well, but its highly unlikely that they will have any considerable impact on the protocol scalability to smaller platforms. They will of course have an impact on the cloud side of things.

**Comment by LorenzMeier** [comment]
I did it for Attitude: 56 bytes Cap'n proto, 36 bytes MAVLink - duh!

**Comment by LorenzMeier** [comment]
https://github.com/mavlink/v2_playground/blob/master/test.cpp

**Comment by LorenzMeier** [comment]
It would be great to include more libs in the test harness and run the attitude message for all of them as a first step.

**Comment by LorenzMeier** [comment]
Turns out Cap'n Proto defaults to unpacked output. In packed variant the message is 40 bytes (vs. 36 MAVLink). I expect that ASN.1 BER would be worse, but we should try it. ASN.1 PER is not extensible.

**Comment by LorenzMeier** [comment]
@mrpollo @liamstask Time to get your hands dirty! How about providing a version packed with Flat buffers and one with ASN.1?

**Comment by LorenzMeier** [comment]
And see if you can recruit someone in the office willing to do a SBE variant - although its not as great for log files https://capnproto.org/news/2014-06-17-capnproto-flatbuffers-sbe.html.

**Comment by LorenzMeier** [comment]
msgpack: {"timestamp": 2384573485, "gyro_speed": {"y": -0.00023, "x": 0.00023, "z": 0.00023}, "orientation": {"y": 0.12, "x": 0.23, "z": 0.09, "w": 0.9}}
is 145 byte, encoded 118

**Comment by LorenzMeier** [comment]
@vooon if you do more testing, please report results here.

**Comment by LorenzMeier** [comment]
flatbuffers
result: 64 / 56 bytes
perhaps struct require less memory than table
https://github.com/mavlink/v2_playground/blob/play_with_subtypes_vooon/mavlink.fbs
yes, it is difference between struct and table: i changed type for Vector3 and Quaternion, result 104 byte for encoded message

**Comment by vooon** [comment]
Result for msgpack on [essential data][1] (hand crafted, some data from logs).

Pretty printed JSON: 550 byte, one-string: 459 byte, msgpack: 343 bytes.

[1]: https://github.com/mavlink/v2_playground/blob/play_with_subtypes_vooon/test_msgpack_essentials.json

**Comment by vooon** [comment]
Result for FlatBuffers on essentials v1 and v2: 136 & 136 bytes.
I suppose that there lots of paddings in wire.

**Comment by vooon** [comment]
Result for Protocol Buffers in essentials: 109 & 111 bytes.

**Comment by vooon** [comment]
Updated issue description with result table.
For msgpack i used python module in ipython:

```python
len(msgpack.dumps(json.load(open('testdata/msgpack_essentials_v2.json'))))
```

**Comment by LorenzMeier** [comment]
@vooon Great work, this has really helped moving things forward!

**Comment by vooon** [comment]
Also i tested [LCM message encoder][1] ([using my vala implementation][2]), it is more for fun but may be useful.

Some thing wrong with subtyped attitude (perhaps JSON loader don't do proper field initialization).
But as i know that serializer it must be exactly same size as normal attitude_t.
Note that each message stores 64-bit hash in front of message.

```
$ ./test_lcm essentials_v2_t ./testdata/msgpack_essentials_v2.json | wc                               
Encoded size: 71
      0       1      71
```

[1]: https://github.com/mavlink/v2_playground/blob/play_with_subtypes_vooon/test_lcm.vala
[2]: https://github.com/vooon/lcm-vala

**Comment by vooon** [comment]
Tested CBOR (RFC-7049) format. It is like MesasgePack. Table updated.

**Comment by LorenzMeier** [comment]
Thanks, much appreciated!

**Comment by liamstask** [comment]
@LorenzMeier - having looked at a translation of [mavlink msgs to capn proto](https://github.com/mavlink/v2_playground/blob/mavlink-capnp-conversion/mavlink-to-capnp/common_annotated.capnp) (thanks @schwa!), I noticed the [Attitude msg def](https://github.com/mavlink/v2_playground/blob/master/mavlink.capnp) doesn't match the current def in the [c_library](https://github.com/mavlink/c_library/blob/master/common/mavlink_msg_attitude.h) - the timestamp uses UInt64 and it has an additional field.

Just wanted to confirm the 36 byte mavlink measurement corresponds to that same layout. Assuming it does, those 36 bytes are exactly the size of the fields in the struct, and don't reflect the size of the serialized mavlink msg (ie, with headers, checksum, etc).

Relatedly, I made a few notes in the commit msg on 9a134223fbec discussing the difference between the size of the fields themselves vs additional format overhead for capn proto in particular. The code in test.cpp at the moment includes the capnp serialization overhead (a la https://capnproto.org/encoding.html#serialization-over-a-stream), so looks like it might not be a direct comparison to the mavlink msg size.

**Comment by vooon** [comment]
@liamstask good catch, so same 40 byte for ATTITUDE with 64-bit timestamp.

---

Leave here notes about RTPS protocol. At September 1-st ROS guys announced alpha1 release.
Most intersecting in topic of mavlink2 is [FreeRTPS project](https://github.com/ros2/freertps).

In short: this project implements RTPS protocol - pub/sub transport of DDS. That protocol implements reliable and unreliable message transfer over UDP/IP (serial transport planned).
That, i think, essential feature for mavlink2: unreliable topics for telemetry (attitude/gps etc) and reliable topics for param/command.

I'm looked into serialization (CDR) used in both CORBA and DDS. And think that it should be similar to protobuf or capnp. But i'm not tested it.

**Comment by schwa** [comment]
@liamstask Any idea why there's a disparity between the two Altitude message representations?

**Comment by schwa** [comment]
I think measuring just encoded sizes of different formats is useful but doesn't cover the whole story. For example even best tuned JSON parser is going to be much, much slower than Capn'Proto (and probably the other formats) when decoding data. Quality of source code generated by schema based encoders is another factor too (Thrift, which isn't on the list produces notoriously awful code).

We should also be able to reject, out of hand, unsuitable formats like ASN.1 which we know are going to be nothing but trouble :-)

**Comment by LorenzMeier** [comment]
@vooon Attitude quaternion has 36 bytes *with* 8 bytes timestamp. So your comment about 40 bytes is not accurate.

**Comment by LorenzMeier** [comment]
@schwa I presume you're talking about cloud performance side of things, as covered here:
https://docs.google.com/document/d/1DzXZ3Ty8BhJZX2-13DoJNz7ERYPc_zjBTPtr4qo_2b4/edit#

**Comment by schwa** [comment]
Well - I'm talking about Mobile performance - which is what I'm focusing on. JSON parsing involves allocating a lot of objects (effectively one for each value in the JSON file - although you can do tricks with keys and tagged pointers), Capn'Proto you load the data into memory and you're done.

**Comment by LorenzMeier** [comment]
Yes, and that's covered in the document above, see the 2nd half of page 1.

**Comment by vooon** [comment]
Done test for OMG CDR (using Fast-CDR library).
There option to select CORBA or DDS encoding, but really it no matters, i see same result.
Table updated.

**Comment by vooon** [comment]
Also note about extensibility: capnp allows only adding new fields to message's bottom, absence of optional fields forbid us from remove deprecated fields from wire. I think CDR may do same thing, but depends on realization.
In that case only protocol buffers may remove deprecated field from wire and save bandwidth.

**Comment by vooon** [comment]
Here is great detailed answer from FreeRTPS guy:  https://github.com/ros2/freertps/issues/16#issuecomment-137831245

**Comment by vooon** [comment]
I added data form #2 for Protocol Buffers v3.

**Comment by vooon** [comment]
I realized that did not test JSON, CBOR, MsgPack with packing just values in Array instead of Map (Object).
In that approach is more like proto/capnp but without field number tag so extensibility has same problems as CDR.

For CBOR i expect almost same size as proto because it uses similar varint scheme.

**Comment by liamstask** [comment]
thanks @vooon - just as a note for any other observers, the difference in size between proto v2 and v3 in the table above is not due to a change in the protoc encoding, but due to a change in the data being encoded. i.e., smaller values require fewer bytes.
